contrib/win32/win32compat/misc.c

```diff
@@ -1046,8 +1048,228 @@ readpassphrase(const char *prompt, char *outBuf, size_t outBufLen, int flags)
+	if (getDomainName(w_machine_domain_name, 64)) {
+		if(!machine_domain_name)
+			machine_domain_name = utf16_to_utf8(w_machine_domain_name);
+
+		wchar_t local_user_fmt[256] = { 0, };
+		if (user_domain) {
+			wcscpy_s(local_user_fmt, 256, w_machine_domain_name);
+			wcscat_s(local_user_fmt, 256, L"\\");
+		}
+
+		wcscat_s(local_user_fmt, 256, w_username);
+		nStatus = NetUserGetLocalGroups(NULL,
+			    local_user_fmt,
+			    0,
+			    flags,
+			    (LPBYTE *)&local_groups,
+			    MAX_PREFERRED_LENGTH,
+			    &num_local_groups_read,
+			    &total_local_groups);
+
+		if (NERR_Success != nStatus)
+			error("Failed to get local groups on this machine, error: %d\n", nStatus);
+	}
+	
+	if (user_domain) {
+		/* Fetch Domain groups */
+		nStatus = NetGetDCName(NULL, w_machine_domain_name, (LPBYTE *)&wszDCName);
+		if (NERR_Success == nStatus) {
+			debug3("domain controller name: %ls", wszDCName);
+
+			nStatus = NetUserGetLocalGroups(wszDCName,
+				    w_username,
+				    0,
+				    flags,
+				    (LPBYTE *)&domain_groups,
+				    MAX_PREFERRED_LENGTH,
+				    &num_domain_groups_read,
+				    &total_domain_groups);
+
+			if (NERR_Success != nStatus)
+				error("Failed to get domain groups from DC:%s error: %d\n", wszDCName, nStatus);
+		}
+		else
+			error("Failed to get the domain controller name, error: %d\n", nStatus);
+
+		/* Fetch global, universal groups */
+		nStatus = NetUserGetGroups(wszDCName,
+			w_username,
+			0,
+			(LPBYTE *)&global_universal_groups,
+			MAX_PREFERRED_LENGTH,
+			&num_global_universal_groups_read,
+			&total_global_universal_groups);
+
+		if (NERR_Success != nStatus)
+			error("Failed to get global,universal groups from DC:%ls error: %d\n", wszDCName, nStatus);
+	}
+
+	int total_user_groups = num_local_groups_read + num_domain_groups_read + num_global_universal_groups_read;
+
+	/* populate the output */
+	char **user_groups = NULL;
+	int num_user_groups = 0;
+	user_groups = malloc(total_user_groups * sizeof(*user_groups));	
+
+	populate_user_groups(user_groups, &num_user_groups, num_local_groups_read, total_local_groups, (LPBYTE) local_groups, LOCAL_GROUP);
+	if (user_domain) {
+		populate_user_groups(user_groups, &num_user_groups, num_domain_groups_read, total_domain_groups, (LPBYTE)domain_groups, DOMAIN_GROUP);
+		populate_user_groups(user_groups, &num_user_groups, num_global_universal_groups_read, total_global_universal_groups, (LPBYTE)global_universal_groups, GLOBAL_UNIVERSAL_GROUP);
+	}
+	
+	for (int i = 0; i < num_user_groups; i++)
+		to_lower_case(user_groups[i]);
+
+	print_user_groups(user, user_groups, num_user_groups);
+
+	if(local_groups)
+		NetApiBufferFree(local_groups);
+
+	if(domain_groups)
+		NetApiBufferFree(domain_groups);
+
+	if(global_universal_groups)
+		NetApiBufferFree(global_universal_groups);
+	
+	*ngroups = num_user_groups;
+	return user_groups;
+}
+
+/* This method will return in "user?group" format */
+char *
+append_domain_to_groupname(char *groupname)
```

Comment:
No need to put "?" here. We resolve a user name in user@domain format. We should resolve a group in group@domain format too for consistency purposes.